src/composed/signed_key/parse.rs

```diff
@@ -9,49 +11,64 @@ use crate::errors::{Error, Result};
+    mut input: R,
 ) -> Result<(
     Box<dyn Iterator<Item = Result<PublicOrSecret>> + 'a>,
-    BTreeMap<String, String>,
+    Option<BTreeMap<String, String>>,
 )> {
-    let mut dearmor = armor::Dearmor::new(input);
-    dearmor.read_header()?;
-    // Safe to unwrap, as read_header succeeded.
-    let typ = dearmor
-        .typ
-        .ok_or_else(|| format_err!("dearmor failed to retrieve armor type"))?;
+    // Peek at the first byte in the reader
+    let first = input.read_u8()?;
+    let _ = input.seek(SeekFrom::Start(0))?;
 
-    // TODO: add typ information to the key possibly?
-    match typ {
-        // Standard PGP types
-        BlockType::PublicKey
-        | BlockType::PrivateKey
-        | BlockType::Message
-        | BlockType::MultiPartMessage(_, _)
-        | BlockType::Signature
-        | BlockType::File => {
-            let headers = dearmor.headers.clone(); // FIXME: avoid clone
-                                                   // TODO: check that the result is what it actually said.
-            Ok((from_bytes_many(dearmor), headers))
-        }
-        BlockType::PublicKeyPKCS1(_)
-        | BlockType::PublicKeyPKCS8
-        | BlockType::PublicKeyOpenssh
-        | BlockType::PrivateKeyPKCS1(_)
-        | BlockType::PrivateKeyPKCS8
-        | BlockType::PrivateKeyOpenssh => {
-            unimplemented_err!("key format {:?}", typ);
+    // If the first bit of the first byte is set, we assume this is binary OpenPGP data, otherwise
+    // we try to parse it as ASCII-armored.
+    let binary = first & 0x80 != 0;
+
+    if !binary {
+        let mut dearmor = armor::Dearmor::new(input);
+        dearmor.read_header()?;
+
+        // Safe to unwrap, as read_header succeeded.
+        let typ = dearmor
+            .typ
+            .ok_or_else(|| format_err!("dearmor failed to retrieve armor type"))?;
+
+        // TODO: add typ information to the key possibly?
+        match typ {
+            // Standard PGP types
+            BlockType::PublicKey
+            | BlockType::PrivateKey
+            | BlockType::Message
+            | BlockType::MultiPartMessage(_, _)
+            | BlockType::Signature
+            | BlockType::File => {
+                let headers = dearmor.headers.clone(); // FIXME: avoid clone
+
+                // TODO: check that the header matches what we expect.
+                Ok((from_bytes_many(dearmor), Some(headers)))
+            }
+            BlockType::PublicKeyPKCS1(_)
+            | BlockType::PublicKeyPKCS8
+            | BlockType::PublicKeyOpenssh
+            | BlockType::PrivateKeyPKCS1(_)
+            | BlockType::PrivateKeyPKCS8
+            | BlockType::PrivateKeyOpenssh => {
+                unimplemented_err!("key format {:?}", typ);
+            }
         }
+    } else {
+        Ok((from_bytes_many(input), None))
     }
 }
 
 /// Parses a list of secret and public keys from raw bytes.
-pub fn from_bytes_many<'a>(
```

Comment:
I think in any case we should still keep explicit versions public, such that places where the format is already known can call those

Reply:
Right, agreed